Index: tif_open.c
===================================================================
--- tif_open.c	(revision 48037)
+++ tif_open.c	(working copy)
@@ -516,6 +516,9 @@
 TIFFSetClientdata(TIFF* tif, thandle_t newvalue)
 {
 	thandle_t m = tif->tif_clientdata;
+#ifdef USE_WIN32_FILEIO
+	newvalue = (thandle_t)_get_osfhandle((int)newvalue);
+#endif /* USE_WIN32_FILEIO */
 	tif->tif_clientdata = newvalue;
 	return m;
 }
 