Index: info.c
===================================================================
--- info.c	(revision 23123)
+++ info.c	(working copy)
@@ -30,6 +30,7 @@
 #include "winbase.h"
 #include "winver.h"
 #include "winternl.h"
+#include "winuser.h"
 #include "wine/winuser16.h"
 #include "wine/unicode.h"
 #include "winerror.h"
@@ -302,6 +303,7 @@
     return len;
 }
 
+#ifndef __REACTOS__
 /***********************************************************************
  *           VERSION_GetFileVersionInfo_16             [internal]
  *
@@ -435,6 +438,7 @@
 
     return len;
 }
+#endif /* ! __REACTOS__ */
 
 /***********************************************************************
  *           GetFileVersionInfoSizeW         [VERSION.@]
@@ -468,6 +471,10 @@
 
     if (!len)
     {
+#ifdef __REACTOS__
+        SetLastError(ERROR_FILE_NOT_FOUND);
+        return 0;
+#else /* __REACTOS__ */
         LPSTR filenameA;
 
         len = WideCharToMultiByte( CP_ACP, 0, filename, -1, NULL, 0, NULL, NULL );
@@ -499,6 +506,7 @@
          * doesn't seem to be a problem (len is bigger than info->wLength).
          */
          len = (len - sizeof(VS_FIXEDFILEINFO)) * 4;
+#endif /* ! __REACTOS__ */
     }
     else
     {
@@ -563,6 +571,10 @@
 
     if (!len)
     {
+#ifdef __REACTOS__
+        SetLastError(ERROR_RESOURCE_DATA_NOT_FOUND);
+        return FALSE;
+#else /* __REACTOS__ */
         LPSTR filenameA;
 
         len = WideCharToMultiByte( CP_ACP, 0, filename, -1, NULL, 0, NULL, NULL );
@@ -578,6 +590,7 @@
             return FALSE;
         }
         /* We have a 16bit resource. */
+#endif /* ! __REACTOS__ */
     }
     else 
     {
Index: resource.c
===================================================================
--- resource.c	(revision 23123)
+++ resource.c	(working copy)
@@ -42,6 +42,9 @@
 
 #include "wine/debug.h"
 
+/* winnt.h */
+#define IMAGE_FILE_RESOURCE_DIRECTORY		2
+
 WINE_DEFAULT_DEBUG_CHANNEL(ver);
 
 
@@ -168,6 +171,7 @@
     return 0;
 }
 
+#ifndef __REACTOS__
 /***********************************************************************
  *           load_ne_resource         [internal]
  */
@@ -266,6 +271,7 @@
     HeapFree( GetProcessHeap(), 0, resTab );
     return TRUE;
 }
+#endif /* ! __REACTOS__ */
 
 /***********************************************************************
  *           load_pe_resource         [internal]
@@ -411,8 +416,13 @@
     switch ( read_xx_header( lzfd ) )
     {
     case IMAGE_OS2_SIGNATURE:
+#ifdef __REACTOS__
+        ERR("OS2 Images not supported under ReactOS at this time.");
+        retv = 0;
+#else
         retv = find_ne_resource( lzfd, lpszResType, lpszResId,
                                  &reslen, lpdwFileOffset );
+#endif
         break;
 
     case IMAGE_NT_SIGNATURE:
@@ -450,8 +460,13 @@
         switch ( read_xx_header( lzfd ) )
         {
         case IMAGE_OS2_SIGNATURE:
+#ifdef __REACTOS__
+            ERR("OS2 Images not supported under ReactOS at this time.");
+            retv = 0;
+#else
             retv = find_ne_resource( lzfd, lpszResType, lpszResId,
                                      &reslen, &dwFileOffset );
+#endif
             break;
 
         case IMAGE_NT_SIGNATURE:
