Index: Makefile.in
===================================================================
RCS file: /home/wine/wine/dlls/icmp/Makefile.in,v
retrieving revision 1.13
diff -u -r1.13 Makefile.in
--- Makefile.in	11 Oct 2003 01:09:19 -0000	1.13
+++ Makefile.in	7 Dec 2004 22:39:27 -0000
@@ -4,6 +4,7 @@
 VPATH     = @srcdir@
 MODULE    = icmp.dll
 IMPORTS   = kernel32
+EXTRALIBS = @SOCKETLIBS@
 
 C_SRCS = icmp_main.c
 
Index: ip.h
===================================================================
RCS file: /home/wine/wine/dlls/icmp/ip.h,v
retrieving revision 1.3
diff -u -r1.3 ip.h
--- ip.h	6 Jan 2004 22:08:34 -0000	1.3
+++ ip.h	7 Dec 2004 22:39:27 -0000
@@ -39,6 +39,21 @@
  */
 #define	IPVERSION	4
 
+/* A little magic to make the Windows build happy. */
+#if defined(_MSC_VER) || defined(__MINGW32__)
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#include <time.h>
+
+typedef u_short n_short;
+typedef u_int n_long;
+typedef u_int n_time;
+
+#define EMSGSIZE WSAEMSGSIZE
+#define ENETUNREACH WSAENETUNREACH
+#define EHOSTUNREACH WSAEHOSTUNREACH
+#endif
+
 /*
  * Structure of an internet header, naked of options.
  */
