diff --git "a/dll/directx/wine/ddraw/ddraw.c" "b/dll/directx/wine/ddraw/ddraw.c"
index e5acf2caff5..2100e137387 100644
--- "a/dll/directx/wine/ddraw/ddraw.c"
+++ "b/dll/directx/wine/ddraw/ddraw.c"
@@ -1,4 +1,4 @@
-/*
+﻿/*
  * Copyright 1997-2000 Marcus Meissner
  * Copyright 1998-2000 Lionel Ulmer
  * Copyright 2000-2001 TransGaming Technologies Inc.
@@ -1132,6 +1132,7 @@ static HRESULT WINAPI ddraw7_SetDisplayMode(IDirectDraw7 *iface, DWORD width, DW
 
     /* TODO: The possible return values from msdn suggest that the screen mode
      * can't be changed if a surface is locked or some drawing is in progress. */
+
     if (SUCCEEDED(hr = wined3d_set_adapter_display_mode(ddraw->wined3d, WINED3DADAPTER_DEFAULT, &mode)))
     {
         if (ddraw->primary)
@@ -1149,6 +1150,10 @@ static HRESULT WINAPI ddraw7_SetDisplayMode(IDirectDraw7 *iface, DWORD width, DW
 
     InterlockedCompareExchange(&ddraw->device_state, DDRAW_DEVICE_STATE_NOT_RESTORED, DDRAW_DEVICE_STATE_OK);
 
+    if (ddraw->cooperative_level & DDSCL_EXCLUSIVE)
+       SetWindowPos(ddraw->dest_window, HWND_TOP, 0, 0, width, height, SWP_SHOWWINDOW | SWP_NOACTIVATE);
+       TRACE("DirectDraw window has been resized\n");
+
     wined3d_mutex_unlock();
 
     switch (hr)
