--- a/dll/win32/kernel32/kernel32.spec
+++ b/dll/win32/kernel32/kernel32.spec
@@ -715,7 +715,11 @@
 @ stdcall -version=0x600+ InitOnceComplete(ptr long ptr)
 @ stdcall -version=0x600+ InitOnceExecuteOnce(ptr ptr ptr ptr)
 @ stdcall -version=0x600+ InitOnceInitialize(ptr) ntdll.RtlRunOnceInitialize
+# FIXME: For amd64, remove version restriction for GCC 13 compatibility
+@ stdcall -arch=amd64 InitializeConditionVariable(ptr)
 @ stdcall -version=0x600+ InitializeConditionVariable(ptr) ntdll.RtlInitializeConditionVariable
 @ stdcall InitializeCriticalSection(ptr)
 @ stdcall InitializeCriticalSectionAndSpinCount(ptr long)
 @ stdcall InitializeCriticalSectionEx(ptr long long)
@@ -1132,8 +1136,12 @@
 @ stdcall SetVolumeMountPointW(wstr wstr)
 @ stdcall SetWaitableTimer(ptr ptr long ptr ptr long)
 @ stdcall SetWaitableTimerEx(ptr ptr long ptr ptr ptr long)
+# FIXME: For amd64, remove version restriction for GCC 13 compatibility
+@ stdcall -arch=amd64 SleepConditionVariableCS(ptr ptr long)
+@ stdcall -arch=amd64 SleepConditionVariableSRW(ptr ptr long long)
 @ stdcall -version=0x600+ SleepConditionVariableCS(ptr ptr long)
 @ stdcall -version=0x600+ SleepConditionVariableSRW(ptr ptr long long)
 @ stdcall SignalObjectAndWait(ptr ptr long long)
 @ stdcall SizeofResource(ptr ptr)
 @ stdcall Sleep(long)
@@ -1208,8 +1216,12 @@
 @ stdcall VirtualQueryEx(ptr ptr ptr long)
 @ stdcall VirtualUnlock(ptr long)
 @ stdcall WTSGetActiveConsoleSessionId()
+# FIXME: For amd64, remove version restriction for GCC 13 compatibility
+@ stdcall -arch=amd64 WakeAllConditionVariable(ptr)
+@ stdcall -arch=amd64 WakeConditionVariable(ptr)
 @ stdcall -version=0x600+ WakeAllConditionVariable(ptr) ntdll.RtlWakeAllConditionVariable
 @ stdcall -version=0x600+ WakeConditionVariable(ptr) ntdll.RtlWakeConditionVariable
 @ stdcall WaitCommEvent(ptr ptr ptr)
 @ stdcall WaitForDebugEvent(ptr long)
 @ stdcall WaitForMultipleObjects(long ptr long long)