Index: include/reactos/exeformat.h
===================================================================
--- include/reactos/exeformat.h	(revision 34037)
+++ include/reactos/exeformat.h	(working copy)
@@ -35,6 +35,7 @@
 typedef NTSTATUS (NTAPI * PEXEFMT_CB_READ_FILE)
 (
  IN PVOID File,
+ ULONG SectorSize,
  IN PLARGE_INTEGER Offset,
  IN ULONG Length,
  OUT PVOID * Data,
Index: ntoskrnl/include/internal/mm.h
===================================================================
--- ntoskrnl/include/internal/mm.h	(revision 34040)
+++ ntoskrnl/include/internal/mm.h	(working copy)
@@ -187,6 +187,9 @@
     ULONG Flags;
     ULONG Characteristics;
     BOOLEAN WriteCopy;
+    LIST_ENTRY ListEntry;
+    ULONG BytesPerSector;
+    PFILE_OBJECT FileObject;
 } MM_SECTION_SEGMENT, *PMM_SECTION_SEGMENT;
 
 typedef struct _MM_IMAGE_SECTION_OBJECT
@@ -203,6 +206,10 @@
     BOOLEAN Executable;
     ULONG NrSegments;
     ULONG ImageSize;
+    ULONG RefCount;
+    LIST_ENTRY ListEntry;
+    PFILE_OBJECT FileObject;
+    ULONG BytesPerSector;
     PMM_SECTION_SEGMENT Segments;
 } MM_IMAGE_SECTION_OBJECT, *PMM_IMAGE_SECTION_OBJECT;
 
@@ -221,6 +228,16 @@
     };
 } ROS_SECTION_OBJECT, *PROS_SECTION_OBJECT;
 
+typedef struct
+{
+   ROS_SECTION_OBJECT* Section;
+   ULONG ViewOffset;
+   LIST_ENTRY ViewListEntry;
+   PMM_SECTION_SEGMENT Segment;
+// BOOLEAN WriteCopyView;
+   LIST_ENTRY RegionListHead;
+} SECTION_DATA, *PSECTION_DATA;
+  
 typedef struct _MEMORY_AREA
 {
     PVOID StartingAddress;
@@ -235,17 +252,10 @@
     ULONG PageOpCount;
     union
     {
+        SECTION_DATA SectionData;
         struct
         {
-            ROS_SECTION_OBJECT* Section;
-            ULONG ViewOffset;
-            PMM_SECTION_SEGMENT Segment;
-            BOOLEAN WriteCopyView;
             LIST_ENTRY RegionListHead;
-        } SectionData;
-        struct
-        {
-            LIST_ENTRY RegionListHead;
         } VirtualMemoryData;
     } Data;
 } MEMORY_AREA, *PMEMORY_AREA;
Index: ntoskrnl/include/internal/cc.h
===================================================================
--- ntoskrnl/include/internal/cc.h	(revision 34040)
+++ ntoskrnl/include/internal/cc.h	(working copy)
@@ -102,19 +102,29 @@
     LONG ActivePrefetches;
 } PFSN_PREFETCHER_GLOBALS, *PPFSN_PREFETCHER_GLOBALS;
 
+#define CACHE_VIEW_SIZE	(128 * 1024) // 128kB
+
+struct _BCB;
+
+typedef struct
+{
+    SECTION_DATA SectionData;
+    PVOID BaseAddress;
+    ULONG RefCount;
+    struct _BCB* Bcb;
+    LIST_ENTRY ListEntry;
+} CACHE_VIEW, *PCACHE_VIEW;
+
 typedef struct _BCB
 {
-    LIST_ENTRY BcbSegmentListHead;
-    LIST_ENTRY BcbRemoveListEntry;
-    BOOLEAN RemoveOnClose;
-    ULONG TimeStamp;
     PFILE_OBJECT FileObject;
-    ULONG CacheSegmentSize;
-    LARGE_INTEGER AllocationSize;
-    LARGE_INTEGER FileSize;
-    PCACHE_MANAGER_CALLBACKS Callbacks;
-    PVOID LazyWriteContext;
-    KSPIN_LOCK BcbLock;
+    CC_FILE_SIZES FileSizes;
+    BOOLEAN PinAccess;
+    PCACHE_MANAGER_CALLBACKS CallBacks;
+    PVOID LazyWriterContext;
+    PCACHE_VIEW CacheView[2048];
+    PVOID LargeCacheView;
+    PROS_SECTION_OBJECT Section;
     ULONG RefCount;
 #if defined(DBG) || defined(KDBG)
 	BOOLEAN Trace; /* enable extra trace output for this BCB and it's cache segments */
@@ -159,8 +169,8 @@
 typedef struct _INTERNAL_BCB
 {
     PUBLIC_BCB PFCB;
-    PCACHE_SEGMENT CacheSegment;
-    BOOLEAN Dirty;
+    PBCB Bcb;
+    ULONG Index;
     CSHORT RefCount; /* (At offset 0x34 on WinNT4) */
 } INTERNAL_BCB, *PINTERNAL_BCB;
 
Index: drivers/filesystems/fastfat/vfat.h
===================================================================
--- drivers/filesystems/fastfat/vfat.h	(revision 34037)
+++ drivers/filesystems/fastfat/vfat.h	(working copy)
@@ -6,7 +6,7 @@
 #ifdef __GNUC__
 #include <ccros.h>
 
-#define USE_ROS_CC_AND_FS
+//#define USE_ROS_CC_AND_FS
 #else
 #define KEBUGCHECK KeBugCheck
 #define KEBUGCHECKEX KeBugCheckEx
