diff --git a/ntoskrnl/mm/ARM3/syspte.c b/ntoskrnl/mm/ARM3/syspte.c
index 87b862c265f..059aebd3a8d 100644
--- a/ntoskrnl/mm/ARM3/syspte.c
+++ b/ntoskrnl/mm/ARM3/syspte.c
@@ -247,7 +247,10 @@ MiReserveSystemPtes(IN ULONG NumberOfPtes,
                     IN MMSYSTEM_PTE_POOL_TYPE SystemPtePoolType)
 {
     PMMPTE PointerPte;
-
+     if( NumberOfPtes > 32768 )
+    {
+        return NULL;
+    }
     //
     // Use the extended function
     //
