Index: gdiplus.c
===================================================================
RCS file: /home/wine/wine/dlls/gdiplus/gdiplus.c,v
retrieving revision 1.15
diff -u -r1.15 gdiplus.c
--- gdiplus.c	8 Jan 2008 16:22:52 -0000	1.15
+++ gdiplus.c	23 Mar 2008 21:21:23 -0000
@@ -200,7 +200,7 @@
     unstretch_angle(&startAngle, x2 / 2.0, y2 / 2.0);
     unstretch_angle(&endAngle, x2 / 2.0, y2 / 2.0);
 
-    count = ceilf(fabs(endAngle - startAngle) / M_PI_2) * 3 + 1;
+    count = ceil(fabs(endAngle - startAngle) / M_PI_2) * 3 + 1;
     /* don't make more than a full circle */
     count = min(MAX_ARC_PTS, count);
 
Index: gdiplus_private.h
===================================================================
RCS file: /home/wine/wine/dlls/gdiplus/gdiplus_private.h,v
retrieving revision 1.58
diff -u -r1.58 gdiplus_private.h
--- gdiplus_private.h	15 Aug 2007 09:36:31 -0000	1.58
+++ gdiplus_private.h	23 Mar 2008 09:25:54 -0000
@@ -46,7 +46,7 @@
 
 static inline INT roundr(REAL x)
 {
-    return (INT) floorf(x + 0.5);
+    return (INT) floor(x + 0.5);
 }
 
 static inline REAL deg2rad(REAL degrees)
Index: graphics.c
===================================================================
RCS file: /home/wine/wine/dlls/gdiplus/graphics.c,v
retrieving revision 1.102
diff -u -r1.102 graphics.c
--- graphics.c	1 Mar 2008 12:15:48 -0000	1.102
+++ graphics.c	23 Mar 2008 21:21:23 -0000
@@ -1464,7 +1464,7 @@
     DeleteObject(SelectObject(graphics->hdc, CreateFontIndirectW(&lfw)));
 
     for(i = 0, j = 0; i < length; i++){
-        if(!isprintW(string[i]) && (string[i] != '\n'))
+        if(!isprint(string[i]) && (string[i] != '\n'))
             continue;
 
         stringdup[j] = string[i];
@@ -1873,7 +1873,7 @@
         nwidth = nheight = INT_MAX;
 
     for(i = 0, j = 0; i < length; i++){
-        if(!isprintW(string[i]) && (string[i] != '\n'))
+        if(!isprint(string[i]) && (string[i] != '\n'))
             continue;
 
         stringdup[j] = string[i];

