Index: activeimmapp.c
===================================================================
--- activeimmapp.c	(Wine c4be28ed)
+++ activeimmapp.c	(working copy)
@@ -664,7 +644,9 @@
 static HRESULT WINAPI ActiveIMMApp_OnDefWindowProc(IActiveIMMApp* This,
         HWND hWnd, UINT Msg, WPARAM wParam, LPARAM lParam, LRESULT *plResult)
 {
+#ifndef __REACTOS__
     FIXME("Stub (%p %x %lx %lx)\n",hWnd,Msg,wParam,lParam);
+#endif
     return E_FAIL;
 }
 
