Index: crypt32.rc
===================================================================
--- crypt32.rc	(revision 22838)
+++ crypt32.rc	(working copy)
@@ -17,12 +17,18 @@
  * License along with this library; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
+
+#define REACTOS_VERSION_DLL
+#define REACTOS_STR_FILE_DESCRIPTION	"CryptoAPI Library\0"
+#define REACTOS_STR_INTERNAL_NAME	"crypt32\0"
+#define REACTOS_STR_ORIGINAL_FILENAME	"crypt32.dll\0"
+
 #include "windef.h"
 #include "winbase.h"
 #include "winuser.h"
 #include "cryptres.h"
 
-#include "version.rc"
+#include <reactos/version.rc>
 
 #include "crypt32_De.rc"
 #include "crypt32_En.rc"
