--- D:/Wine-CVS/wine/dlls/crypt32/rootstore.c	Sat Feb 16 22:49:56 2008
+++ D:/ReactOS-Trunk/reactos/dll/win32/crypt32/rootstore.c	Sat May 10 20:30:25 2008
@@ -317,7 +317,7 @@
     DIR *dir;
 
     TRACE("(%s, %p)\n", debugstr_a(path), store);
-
+    /* UNIX functions = bad for reactos
     dir = opendir(path);
     if (dir)
     {
@@ -340,6 +340,7 @@
             CryptMemFree(filebuf);
         }
     }
+    */
     return ret;
 }
 
