Index: hhctrl.ocx.rbuild
===================================================================
--- hhctrl.ocx.rbuild	(revision 27782)
+++ hhctrl.ocx.rbuild	(working copy)
@@ -5,9 +5,11 @@
 	<importlibrary definition="hhctrl.ocx.spec.def" />
 	<include base="hhctrl">.</include>
 	<include base="ReactOS">include/reactos/wine</include>
+	<include base="ReactOS" root="intermediate">include/reactos</include>
 	<define name="__WINESRC__" />
 	<define name="WINVER">0x600</define>
 	<define name="_WIN32_WINNT">0x600</define>
+	<dependency>wineheaders</dependency>
 	<library>wine</library>
 	<library>advapi32</library>
 	<library>comctl32</library>
