Index: graphics.c
===================================================================
--- graphics.c  (revision 57206)
+++ graphics.c  (working copy)
@@ -2819,6 +2819,7 @@
     ReleaseDC(hwnd, hdc);
 }

+#if CORE_6659_IS_FIXED
 static void test_string_functions(void)
 {
     GpStatus status;
@@ -3138,6 +3139,7 @@

     ReleaseDC(hwnd, hdc);
 }
+#endif /* CORE_6659_IS_FIXED */

 static void test_get_set_interpolation(void)
 {
@@ -3361,7 +3363,9 @@
     test_clear();
     test_textcontrast();
     test_fromMemoryBitmap();
+#if CORE_6659_IS_FIXED
     test_string_functions();
+#endif
     test_get_set_interpolation();
     test_get_set_textrenderinghint();
     test_getdc_scaled();

Index: font.c
===================================================================
--- font.c      (revision 57301)
+++ font.c      (working copy)
@@ -87,6 +87,7 @@
     GdipDeleteFontFamily(fontfamily);
 }

+#if CORE_6660_IS_FIXED
 static void test_logfont(void)
 {
     LOGFONTA lfa, lfa2;
@@ -274,6 +275,7 @@

     GdipDeleteFontFamily(clonedFontFamily);
 }
+#endif // CORE_6660_IS_FIXED

 static void test_fontfamily_properties (void)
 {
@@ -740,8 +742,10 @@
     test_font_metrics();
 #endif
     test_createfont();
+#if CORE_6660_IS_FIXED
     test_logfont();
     test_fontfamily();
+#endif
     test_fontfamily_properties();
     test_getgenerics();
     test_installedfonts();
