Index: Makefile.in
===================================================================
RCS file: /home/wine/wine/dlls/mpr/Makefile.in,v
retrieving revision 1.15
diff -u -r1.15 Makefile.in
--- Makefile.in	5 Apr 2004 23:14:26 -0000	1.15
+++ Makefile.in	16 Oct 2004 16:50:33 -0000
@@ -15,8 +15,7 @@
 	wnet.c
 
 RC_SRCS = \
-	mpr.rc \
-	version.rc
+	mpr.rc
 
 
 @MAKE_DLL_RULES@
Index: mpr.rc
===================================================================
RCS file: /home/wine/wine/dlls/mpr/mpr.rc,v
retrieving revision 1.8
diff -u -r1.8 mpr.rc
--- mpr.rc	9 Sep 2004 19:18:58 -0000	1.8
+++ mpr.rc	16 Oct 2004 16:50:33 -0000
@@ -21,6 +21,8 @@
 #include "winbase.h"
 #include "mprres.h"
 
+#include "version.rc"
+
 #include "mpr_Cs.rc"
 #include "mpr_De.rc"
 #include "mpr_En.rc"
Index: wnet.c
===================================================================
RCS file: /home/wine/wine/dlls/mpr/wnet.c,v
retrieving revision 1.17
diff -u -r1.17 wnet.c
--- wnet.c	30 Jun 2004 18:17:07 -0000	1.17
+++ wnet.c	16 Oct 2004 16:50:34 -0000
@@ -22,6 +22,7 @@
 #include <stdarg.h>
 #include "windef.h"
 #include "winbase.h"
+#include "winuser.h"
 #include "winnls.h"
 #include "winnetwk.h"
 #include "npapi.h"
