Index: usp10.c
===================================================================
--- usp10.c	(revision 25776)
+++ usp10.c	(working copy)
@@ -35,6 +35,8 @@
 #include "wine/debug.h"
 #include "wine/unicode.h"
 
+DWORD STDCALL GetGlyphIndicesW( HDC hdc, LPCWSTR lpstr, int c, LPWORD pgi, DWORD fl);
+BOOL  STDCALL GetCharABCWidthsI(HDC hdc, UINT giFirst, UINT cgi, LPWORD pgi, LPABC lpabc);
 /**
  * some documentation here:
  *   http://www.microsoft.com/typography/developers/uniscribe/uniscribe.htm
@@ -1241,7 +1243,7 @@
     TRACE("\n");
 
     if  (!psa->fNoGlyphIndex) {                                         /* Glyph translate */
-        if (!(GetGlyphIndicesW(get_cache_hdc(psc), pwcChars, cChars, pwOutGlyphs, 0)))
+        if (!(GetGlyphIndicesW((HDC)get_cache_hdc(psc), (LPCWSTR) pwcChars, (int)cChars, (LPWORD) pwOutGlyphs,(DWORD) 0)))
             return S_FALSE;
 
         TRACE("After:  ");
