Index: usp10.c
===================================================================
--- usp10.c     (revision 54504)
+++ usp10.c     (working copy)
@@ -1989,3 +3159,9 @@
     for (i = 0; i < num_glyphs; i++) justify[i] = advance[i];
     return S_OK;
 }
+
+BOOL gbLpkPresent = FALSE;
+VOID WINAPI LpkPresent()
+{
+    gbLpkPresent = TRUE; /* Turn it on this way! Wine is out of control! */
+}
Index: usp10.spec
===================================================================
--- usp10.spec	(revision 54504)
+++ usp10.spec	(working copy)
@@ -1,4 +1,4 @@
-@ stub LpkPresent
+@ stdcall LpkPresent()
 @ stdcall ScriptApplyDigitSubstitution(ptr ptr ptr)
 @ stdcall ScriptApplyLogicalWidth(ptr long long ptr ptr ptr ptr ptr ptr)
 @ stdcall ScriptBreak(ptr long ptr ptr)
