Index: usp10.c
===================================================================
--- usp10.c	(revision 34385)
+++ usp10.c	(working copy)
@@ -1726,3 +1726,10 @@
     for (i = 0; i < nbchars; i++) widths[i] = glyph_width[i];
     return S_OK;
 }
+
+
+VOID WINAPI LpkPresent()   
+{   
+    /* FIXME */   
+    ERR("LPK: %s is unimplemented, please try again later.\n", __FUNCTION__);   
+} 
Index: usp10.spec
===================================================================
--- usp10.spec	(revision 34385)
+++ usp10.spec	(working copy)
@@ -1,4 +1,4 @@
-@ stub LpkPresent
+@ stdcall LpkPresent()
 @ stdcall ScriptApplyDigitSubstitution(ptr ptr ptr)
 @ stub ScriptApplyLogicalWidth
 @ stdcall ScriptBreak(ptr long ptr ptr)
