Index: oleaut32_Gr.rc
===================================================================
--- oleaut32_Gr.rc	(revision 23782)
+++ oleaut32_Gr.rc	(working copy)
@@ -0,0 +1,31 @@
+/*
+ * Greek resources for oleaut32
+ *
+ * Copyright 2006 Dj Apal®
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+LANGUAGE LANG_GREEK, SUBLANG_DEFAULT
+
+STRINGTABLE DISCARDABLE
+{
+  IDS_TRUE  "Αληθές"
+  IDS_FALSE "Ψευδές"
+  IDS_YES   "Ναι"
+  IDS_NO    "Όχι"
+  IDS_ON    "Ανοιχτό"
+  IDS_OFF   "Κλειστό"
+}
Index: oleaut32_Ja.rc
===================================================================
--- oleaut32_Ja.rc	(revision 23782)
+++ oleaut32_Ja.rc	(working copy)
@@ -0,0 +1,11 @@
+LANGUAGE LANG_JAPANESE, SUBLANG_DEFAULT
+
+STRINGTABLE DISCARDABLE
+{
+  IDS_TRUE  "True"
+  IDS_FALSE "False"
+  IDS_YES   "‚Ν‚Ά"
+  IDS_NO    "‚Ά‚Ά‚¦"
+  IDS_ON    "ƒIƒ“"
+  IDS_OFF   "ƒIƒt"
+}
Index: oleaut32_Uk.rc
===================================================================
--- oleaut32_Uk.rc	(revision 23782)
+++ oleaut32_Uk.rc	(working copy)
@@ -0,0 +1,31 @@
+/*
+ * Ukrainian resources for oleaut32
+ *
+ * Copyright 2006 Artem Reznikov
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+LANGUAGE LANG_UKRAINIAN, SUBLANG_DEFAULT
+
+STRINGTABLE DISCARDABLE
+{
+  IDS_TRUE  "²ρςθνΰ"
+  IDS_FALSE "Νεοπΰβδΰ"
+  IDS_YES   "Òΰκ"
+  IDS_NO    "Ν³"
+  IDS_ON    "Ββ³μκνενξ"
+  IDS_OFF   "Βθμκνενξ"
+}
Index: oleaut32.rc
===================================================================
--- oleaut32.rc	(revision 23782)
+++ oleaut32.rc	(working copy)
@@ -32,8 +32,10 @@
 #include "oleaut32_Es.rc"
 #include "oleaut32_Cz.rc"
 #include "oleaut32_Fr.rc"
+#include "oleaut32_Gr.rc"
 #include "oleaut32_Hu.rc"
 #include "oleaut32_It.rc"
+#include "oleaut32_Ja.rc"
 #include "oleaut32_Ko.rc"
 #include "oleaut32_Nl.rc"
 #include "oleaut32_No.rc"
@@ -43,6 +45,7 @@
 #include "oleaut32_Sv.rc"
 #include "oleaut32_Th.rc"
 #include "oleaut32_Tr.rc"
+#include "oleaut32_Uk.rc"
 
 /*
  * FIXME:
Index: oleaut32.rbuild
===================================================================
--- oleaut32.rbuild	(revision 23782)
+++ oleaut32.rbuild	(working copy)
@@ -20,6 +20,7 @@
 	<library>comctl32</library>
 	<library>urlmon</library>
 	<library>uuid</library>
+	<library>pseh</library>
 	<file>connpt.c</file>
 	<file>dispatch.c</file>
 	<file>hash.c</file>
