Index: reader.c
===================================================================
RCS file: /home/wine/wine/dlls/richedit/reader.c,v
retrieving revision 1.13
diff -u -r1.13 reader.c
--- reader.c	30 Jan 2004 22:56:33 -0000	1.13
+++ reader.c	17 Feb 2004 23:16:04 -0000
@@ -126,11 +126,17 @@
     {
         char buff[10];
         long pcb;
-        info->editstream.pfnCallback(info->editstream.dwCookie, buff, 1, &pcb);
+        info->editstream.pfnCallback(info->editstream.dwCookie, buff, sizeof(buff), &pcb);
         if(pcb == 0)
            return EOF;
         else
-           CHARLIST_Enqueue(&info->inputCharList, buff[0]);
+        {
+           int i;
+           for (i = 0; i < pcb; i++)
+           {
+               CHARLIST_Enqueue(&info->inputCharList, buff[i]);
+           }
+        }
     }
     myChar = CHARLIST_Dequeue(&info->inputCharList);
     return (int) myChar;
