Index: dpa.c
===================================================================
--- dpa.c	(revision 25766)
+++ dpa.c	(working copy)
@@ -25,6 +25,7 @@
 
 #include "windef.h"
 #include "winbase.h"
+#include "wingdi.h"
 #include "winuser.h"
 #include "commctrl.h"
 #include "objidl.h"
Index: monthcal.c
===================================================================
--- monthcal.c	(revision 25766)
+++ monthcal.c	(working copy)
@@ -23,6 +23,7 @@
 
 #include "windef.h"
 #include "winbase.h"
+#include "wingdi.h"
 #include "winuser.h"
 
 #include "commctrl.h"
