Index: path.c
===================================================================
RCS file: /home/wine/wine/dlls/shlwapi/path.c,v
retrieving revision 1.41
diff -u -r1.41 path.c
--- path.c	23 Jan 2004 22:45:25 -0000	1.41
+++ path.c	28 Jan 2004 22:02:59 -0000
@@ -32,6 +32,7 @@
 #include "wingdi.h"
 #include "winuser.h"
 #include "winreg.h"
+#include "winnls.h"
 #define NO_SHLWAPI_STREAM
 #include "shlwapi.h"
 #include "wine/debug.h"
Index: string.c
===================================================================
RCS file: /home/wine/wine/dlls/shlwapi/string.c,v
retrieving revision 1.40
diff -u -r1.40 string.c
--- string.c	23 Jan 2004 22:45:25 -0000	1.40
+++ string.c	28 Jan 2004 22:03:00 -0000
@@ -556,7 +556,7 @@
 {
   TRACE("(%s,%s)\n", debugstr_w(lpszStr), debugstr_w(lpszSearch));
 
-  return SHLWAPI_StrStrHelperW(lpszStr, lpszSearch, strncmpW);
+  return SHLWAPI_StrStrHelperW(lpszStr, lpszSearch, (int (*)(LPCWSTR,LPCWSTR,int))wcsncmp);
 }
 
 /*************************************************************************
@@ -665,7 +665,7 @@
 {
   TRACE("(%s,%s)\n", debugstr_w(lpszStr), debugstr_w(lpszSearch));
 
-  return SHLWAPI_StrStrHelperW(lpszStr, lpszSearch, strncmpiW);
+  return SHLWAPI_StrStrHelperW(lpszStr, lpszSearch, (int (*)(LPCWSTR,LPCWSTR,int))_wcsnicmp);
 }
 
 /*************************************************************************
Index: url.c
===================================================================
RCS file: /home/wine/wine/dlls/shlwapi/url.c,v
retrieving revision 1.27
diff -u -r1.27 url.c
--- url.c	23 Jan 2004 22:45:25 -0000	1.27
+++ url.c	28 Jan 2004 22:03:02 -0000
@@ -1396,8 +1396,8 @@
  *  Success: TRUE. lpDest is filled with the computed hash value.
  *  Failure: FALSE, if any argument is invalid.
  */
-HRESULT WINAPI HashData(const unsigned char *lpSrc, DWORD nSrcLen,
-                     unsigned char *lpDest, DWORD nDestLen)
+HRESULT WINAPI HashData(LPBYTE lpSrc, DWORD nSrcLen,
+                     LPBYTE lpDest, DWORD nDestLen)
 {
   INT srcCount = nSrcLen - 1, destCount = nDestLen - 1;
 
