Index: avifile_private.h
===================================================================
--- avifile_private.h	(revision 23782)
+++ avifile_private.h	(working copy)
@@ -23,6 +23,10 @@
 #define MAX_AVISTREAMS 8
 #endif
 
+#ifndef comptypeDIB
+#define comptypeDIB  mmioFOURCC('D','I','B',' ')
+#endif
+
 #ifndef DIBWIDTHBYTES
 #define WIDTHBYTES(i)     (((i+31)&(~31))/8)
 #define DIBWIDTHBYTES(bi) WIDTHBYTES((bi).biWidth * (bi).biBitCount)
