diff -u wine-1.3.4/tools/widl/hash.c tools/widl/hash.c
--- wine-1.3.4/tools/widl/hash.c	2010-09-19 17:48:47.640625000 +0200
+++ tools/widl/hash.c	2010-09-19 19:17:19.000000000 +0200
@@ -21,9 +21,7 @@
 #include <stdio.h>
 #include <stdarg.h>
 
-#include "windef.h"
-#include "winbase.h"
-#include "winnls.h"
+#include <host/nls.h>
 
 #include "widltypes.h"
 #include "hash.h"
diff -u wine-1.3.4/tools/widl/header.c tools/widl/header.c
--- wine-1.3.4/tools/widl/header.c	2010-09-19 17:48:47.640625000 +0200
+++ tools/widl/header.c	2010-10-26 18:30:19.000000000 +0200
@@ -1081,15 +1081,7 @@
 {
   unsigned int ver = get_attrv(iface->attrs, ATTR_VERSION);
   const char *var = get_attrp(iface->attrs, ATTR_IMPLICIT_HANDLE);
-  static int allocate_written = 0;
 
-  if (!allocate_written)
-  {
-    allocate_written = 1;
-    fprintf(header, "void * __RPC_USER MIDL_user_allocate(SIZE_T);\n");
-    fprintf(header, "void __RPC_USER MIDL_user_free(void *);\n\n");
-  }
-
   fprintf(header, "/*****************************************************************************\n");
   fprintf(header, " * %s interface (v%d.%d)\n", iface->name, MAJORVERSION(ver), MINORVERSION(ver));
   fprintf(header, " */\n");
diff -u wine-1.3.4/tools/widl/parser.y tools/widl/parser.y
--- wine-1.3.4/tools/widl/parser.y	2010-09-19 17:49:40.578125000 +0200
+++ tools/widl/parser.y	2010-10-03 16:44:18.781250000 +0200
@@ -1816,6 +1816,12 @@
   else if (is_attr(attrs, ATTR_UUID) && !is_attr(attrs, ATTR_PUBLIC))
     attrs = append_attr( attrs, make_attr(ATTR_PUBLIC) );
 
+  /* Append the SWITCHTYPE attribute to a union if it does not already have one.  */
+  if (type_get_type_detect_alias(type) == TYPE_UNION &&
+      is_attr(attrs, ATTR_SWITCHTYPE) &&
+      !is_attr(type->attrs, ATTR_SWITCHTYPE))
+    type->attrs = append_attr(type->attrs, make_attrp(ATTR_SWITCHTYPE, get_attrp(attrs, ATTR_SWITCHTYPE)));
+
   LIST_FOR_EACH_ENTRY( decl, decls, const declarator_t, entry )
   {
 
diff -u wine-1.3.4/tools/widl/typelib.c tools/widl/typelib.c
--- wine-1.3.4/tools/widl/typelib.c	2010-09-19 17:50:24.000000000 +0200
+++ tools/widl/typelib.c	2010-09-26 20:23:47.000000000 +0200
@@ -35,8 +35,7 @@
 #define NONAMELESSUNION
 #define NONAMELESSSTRUCT
 
-#include "windef.h"
-#include "winbase.h"
+#include <host/typedefs.h>
 
 #include "widl.h"
 #include "utils.h"
diff -u wine-1.3.4/tools/widl/typelib_struct.h tools/widl/typelib_struct.h
--- wine-1.3.4/tools/widl/typelib_struct.h	2010-09-19 17:50:40.953125000 +0200
+++ tools/widl/typelib_struct.h	2010-10-10 00:50:32.921875000 +0200
@@ -302,7 +302,7 @@
  *
  */
 
-#include "pshpack1.h"
+#include <host/pshpack1.h>
 
 typedef struct {
 /*00*/	DWORD SLTG_magic;	/* 0x47544c53  == "SLTG" */
@@ -599,7 +599,7 @@
 WORD typeofarray
 */
 
-#include "poppack.h"
+#include <host/poppack.h>
 
 /*---------------------------END--------------------------------------------*/
 #endif
diff -u wine-1.3.4/tools/widl/widltypes.h tools/widl/widltypes.h
--- wine-1.3.4/tools/widl/widltypes.h	2010-09-19 17:51:38.890625000 +0200
+++ tools/widl/widltypes.h	2010-09-19 19:17:19.656250000 +0200
@@ -21,6 +21,15 @@
 #ifndef __WIDL_WIDLTYPES_H
 #define __WIDL_WIDLTYPES_H
 
+#define S_OK           0
+#define S_FALSE        1
+#define E_OUTOFMEMORY  ((HRESULT)0x8007000EL)
+#define TYPE_E_IOERROR ((HRESULT)0x80028CA2L)
+
+#ifndef max
+#define max(a, b) ((a) > (b) ? a : b)
+#endif
+
 #include <stdarg.h>
 #include <assert.h>
 #include "guiddef.h"
@@ -32,7 +39,9 @@
 typedef GUID UUID;
 #endif
 
+#ifndef TRUE
 #define TRUE 1
+#endif
 #define FALSE 0
 
 typedef struct _loc_info_t loc_info_t;
diff -u wine-1.3.4/tools/widl/write_msft.c tools/widl/write_msft.c
--- wine-1.3.4/tools/widl/write_msft.c	2010-09-19 17:51:48.531250000 +0200
+++ tools/widl/write_msft.c	2010-09-26 20:23:47.000000000 +0200
@@ -40,10 +40,8 @@
 #define NONAMELESSUNION
 #define NONAMELESSSTRUCT
 
-#include "winerror.h"
-#include "windef.h"
-#include "winbase.h"
-#include "winnls.h"
+#include <host/typedefs.h>
+#include <host/nls.h>
 
 #include "widl.h"
 #include "typelib.h"
@@ -2023,7 +2023,10 @@
     }
 
     if (is_attr(interface->attrs, ATTR_DISPINTERFACE))
-        return add_dispinterface_typeinfo(typelib, interface);
+    {
+        add_dispinterface_typeinfo(typelib, interface);
+        return;
+    }
 
     /* midl adds the parent interface first, unless the parent itself
        has no parent (i.e. it stops before IUnknown). */
diff -u wine-1.3.4/tools/widl/typegen.c tools/widl/typegen.c
--- wine-1.3.4/tools/widl/typegen.c	2010-09-19 17:51:48.531250000 +0200
+++ tools/widl/typegen.c	2012-08-12 14:19:47.000000000 +0200
@@ -2345,7 +2345,8 @@
             return;
         }
     }
-    return write_member_type(file, type, cont_is_complex, NULL, elem, NULL, tfsoff);
+    write_member_type(file, type, cont_is_complex, NULL, elem, NULL, tfsoff);
+    return;
 }
 
 static void write_end(FILE *file, unsigned int *tfsoff)
