Index: server.c
===================================================================
--- server.c	(revision 38883)
+++ server.c	(working copy)
@@ -352,12 +352,6 @@
 }
 
 int
-s_sum_parr(int *a[3])
-{
-  return s_sum_pcarr(a, 3);
-}
-
-int
 s_sum_pcarr(int *a[], int n)
 {
   int i, s = 0;
@@ -367,6 +361,12 @@
 }
 
 int
+s_sum_parr(int *a[3])
+{
+  return s_sum_pcarr(a, 3);
+}
+
+int
 s_enum_ord(e_t e)
 {
   switch (e)
