Index: chain.c
===================================================================
--- chain.c	(revision 38919)
+++ chain.c	(working copy)
@@ -205,7 +205,7 @@
 
         config.cbSize = sizeof(config);
         CertCreateCertificateChainEngine(&config, &engine);
-        InterlockedCompareExchangePointer(&CRYPT_defaultChainEngine, engine,
+        (void)InterlockedCompareExchangePointer(&CRYPT_defaultChainEngine, engine,
          NULL);
         if (CRYPT_defaultChainEngine != engine)
             CertFreeCertificateChainEngine(engine);
Index: main.c
===================================================================
--- main.c	(revision 38919)
+++ main.c	(working copy)
@@ -66,7 +66,7 @@
 
         CryptAcquireContextW(&prov, NULL, MS_ENHANCED_PROV_W, PROV_RSA_FULL,
          CRYPT_VERIFYCONTEXT);
-        InterlockedCompareExchangePointer((PVOID *)&hDefProv, (PVOID)prov,
+        (void)InterlockedCompareExchangePointer((PVOID *)&hDefProv, (PVOID)prov,
          NULL);
         if (hDefProv != prov)
             CryptReleaseContext(prov, 0);
Index: rootstore.c
===================================================================
--- rootstore.c	(revision 38919)
+++ rootstore.c	(working copy)
@@ -759,7 +759,7 @@
     {
         HCERTSTORE root = create_root_store();
 
-        InterlockedCompareExchangePointer((PVOID *)&CRYPT_rootStore, root,
+        (void)InterlockedCompareExchangePointer((PVOID *)&CRYPT_rootStore, root,
          NULL);
         if (CRYPT_rootStore != root)
             CertCloseStore(root, 0);
