Index: handle.h
===================================================================
--- handle.h	(revision 30501)
+++ handle.h	(working copy)
@@ -48,6 +48,10 @@
     unsigned int iNextFree;
 } HANDLETABLEENTRY;
 
+/* Prevent conflict with wingdi.h */
+#define tagHANDLETABLE tagHANDLETABLE_RSA
+#define HANDLETABLE HANDLETABLE_RSA
+
 typedef struct tagHANDLETABLE
 {
     unsigned int     iEntries;
Index: rsaenh.c
===================================================================
--- rsaenh.c	(revision 30501)
+++ rsaenh.c	(working copy)
@@ -33,9 +33,9 @@
 #include "winbase.h"
 #include "winreg.h"
 #include "wincrypt.h"
+#include "objbase.h"
 #include "handle.h"
 #include "implglue.h"
-#include "objbase.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(crypt);
 
