Index: dwmapi.spec
===================================================================
--- dwmapi.spec	(revision 49877)
+++ dwmapi.spec	(working copy)
@@ -1,24 +1,24 @@
-100 stub @
-101 stub @
+100 stub DWMAPI_100
+101 stub DWMAPI_101
 102 stdcall DwmEnableComposition (long)
-103 stub @
-104 stub @
-105 stub @
-106 stub @
-107 stub @
-108 stub @
-109 stub @
-110 stub @
-111 stub @
-112 stub @
-113 stub @
+103 stub DWMAPI_103
+104 stub DWMAPI_104
+105 stub DWMAPI_105
+106 stub DWMAPI_106
+107 stub DWMAPI_107
+108 stub DWMAPI_108
+109 stub DWMAPI_109
+110 stub DWMAPI_110
+111 stub DWMAPI_111
+112 stub DWMAPI_112
+113 stub DWMAPI_113
 
-115 stub @
-116 stub @
-117 stub @
-118 stub @
-119 stub @
-120 stub @
+115 stub DWMAPI_114
+116 stub DWMAPI_115
+117 stub DWMAPI_116
+118 stub DWMAPI_117
+119 stub DWMAPI_118
+120 stub DWMAPI_120
 
 @ stub DwmAttachMilContent
 @ stub DwmDefWindowProc
