Index: action.c
===================================================================
RCS file: /home/wine/wine/dlls/msi/action.c,v
retrieving revision 1.40
diff -u -r1.40 action.c
--- action.c	3 Nov 2004 22:16:53 -0000	1.40
+++ action.c	7 Dec 2004 22:53:54 -0000
@@ -28,7 +28,7 @@
 
 #include <stdarg.h>
 #include <stdio.h>
-
+#include <fcntl.h>
 #define COBJMACROS
 
 #include "windef.h"
@@ -39,7 +39,7 @@
 #include "fdi.h"
 #include "msi.h"
 #include "msiquery.h"
-#include "msvcrt/fcntl.h"
+//#include "msvcrt/fcntl.h"
 #include "objbase.h"
 #include "objidl.h"
 #include "msipriv.h"
@@ -2263,7 +2263,7 @@
                     version = HeapAlloc(GetProcessHeap(),0,versize);
                     GetFileVersionInfoW(file->TargetPath, 0, versize, version);
 
-                    VerQueryValueW(version, name, (LPVOID*)&lpVer, &sz);
+                    VerQueryValueW(version, (LPWSTR) name, (LPVOID*)&lpVer, &sz);
 
                     sprintfW(filever,name_fmt,
                         HIWORD(lpVer->dwFileVersionMS),
Index: msi.c
===================================================================
RCS file: /home/wine/wine/dlls/msi/msi.c,v
retrieving revision 1.42
diff -u -r1.42 msi.c
--- msi.c	7 Oct 2004 03:06:50 -0000	1.42
+++ msi.c	7 Dec 2004 22:53:55 -0000
@@ -40,6 +40,9 @@
 
 #include "initguid.h"
 
+UINT WINAPI MsiGetFileVersionW(LPCWSTR szFilePath, LPWSTR lpVersionBuf, DWORD* pcchVersionBuf, LPWSTR lpLangBuf, DWORD* pcchLangBuf);
+
+
 WINE_DEFAULT_DEBUG_CHANNEL(msi);
 
 /*
@@ -1392,7 +1395,7 @@
           lpVersionBuf, pcchVersionBuf?*pcchVersionBuf:0,
           lpLangBuf, pcchLangBuf?*pcchLangBuf:0);
 
-    dwVerLen = GetFileVersionInfoSizeW(szFilePath, NULL);
+    dwVerLen = GetFileVersionInfoSizeW( (LPWSTR) szFilePath, NULL);
     if(!dwVerLen)
         return GetLastError();
 
@@ -1402,12 +1405,12 @@
         goto end;
     }
 
-    if(!GetFileVersionInfoW(szFilePath, 0, dwVerLen, lpVer)) {
+    if(!GetFileVersionInfoW((LPWSTR) szFilePath, 0, dwVerLen, lpVer)) {
         ret = GetLastError();
         goto end;
     }
     if(lpVersionBuf && pcchVersionBuf && *pcchVersionBuf) {
-        if(VerQueryValueW(lpVer, szVersionResource, (LPVOID*)&ffi, &puLen) && puLen > 0) {
+        if(VerQueryValueW(lpVer, (LPWSTR) szVersionResource, (LPVOID*)&ffi, &puLen) && puLen > 0) {
             wsprintfW(tmp, szVersionFormat, HIWORD(ffi->dwFileVersionMS), LOWORD(ffi->dwFileVersionMS), HIWORD(ffi->dwFileVersionLS), LOWORD(ffi->dwFileVersionLS));
             lstrcpynW(lpVersionBuf, tmp, *pcchVersionBuf);
             *pcchVersionBuf = strlenW(lpVersionBuf);
Index: suminfo.c
===================================================================
RCS file: /home/wine/wine/dlls/msi/suminfo.c,v
retrieving revision 1.9
diff -u -r1.9 suminfo.c
--- suminfo.c	7 Oct 2004 03:06:50 -0000	1.9
+++ suminfo.c	7 Dec 2004 22:53:55 -0000
@@ -23,6 +23,8 @@
 #define COBJMACROS
 #define NONAMELESSUNION
 
+#define PRSPEC_PROPID (1)
+
 #include "windef.h"
 #include "winbase.h"
 #include "winreg.h"
@@ -116,6 +118,7 @@
     ERR("storage = %p propertysetstorage = %p\n", db->storage, psstg);
 
     grfMode = STGM_READ | STGM_SHARE_EXCLUSIVE;
+
     r = IPropertySetStorage_Open( psstg, &FMTID_SummaryInformation, grfMode, &ps );
     if( FAILED( r ) )
     {
