Index: string.c
===================================================================
RCS file: /home/wine/wine/libs/unicode/string.c,v
retrieving revision 1.4
diff -u -r1.4 string.c
--- string.c	22 Nov 2003 00:00:53 -0000	1.4
+++ string.c	2 Jan 2004 19:09:14 -0000
@@ -23,6 +23,20 @@
 
 #include "wine/unicode.h"
 
+#undef strlenW
+#undef strcpyW
+#undef strcatW
+#undef strstrW
+#undef strtolW
+#undef strchrW
+#undef strrchrW
+#undef strncmpW
+#undef strncpyW
+#undef strcmpW
+#undef strcmpiW
+#undef strncmpiW
+#undef sprintfW
+
 int strcmpiW( const WCHAR *str1, const WCHAR *str2 )
 {
     for (;;)
