Index: printdlg.c
===================================================================
--- printdlg.c	(revision 23123)
+++ printdlg.c	(working copy)
@@ -352,7 +352,7 @@
 	        lpdm->dmCollate =
 		  (IsDlgButtonChecked(hDlg, chx2) == BST_CHECKED);
 	    if (lpdm->dmFields & DM_COPIES)
-	        lpdm->dmCopies = GetDlgItemInt(hDlg, edt3, NULL, FALSE);
+	        lpdm->u.s.dmCopies = GetDlgItemInt(hDlg, edt3, NULL, FALSE);
 	} else {
 	    if (IsDlgButtonChecked(hDlg, chx2) == BST_CHECKED)
 	        lppd->Flags |= PD_COLLATE;
@@ -435,7 +435,7 @@
 	        lpdm->dmCollate =
 		  (IsDlgButtonChecked(hDlg, chx2) == BST_CHECKED);
 	    if (lpdm->dmFields & DM_COPIES)
-	        lpdm->dmCopies = GetDlgItemInt(hDlg, edt3, NULL, FALSE);
+	        lpdm->u.s.dmCopies = GetDlgItemInt(hDlg, edt3, NULL, FALSE);
 	} else {
 	    if (IsDlgButtonChecked(hDlg, chx2) == BST_CHECKED)
 	        lppd->Flags |= PD_COLLATE;
@@ -600,9 +600,9 @@
                                       Sel, 0);
         if (dm) {
             if (nIDComboBox == cmb2)
-                dm->u1.s1.dmPaperSize = oldWord;
+                dm->u.s.dmPaperSize = oldWord;
             else
-                dm->dmDefaultSource = oldWord;
+                dm->u.s.dmDefaultSource = oldWord;
         }
     }
     else {
@@ -612,9 +612,9 @@
          */
         if (dm) {
             if (nIDComboBox == cmb2)
-                oldWord = dm->u1.s1.dmPaperSize;
+                oldWord = dm->u.s.dmPaperSize;
             else
-                oldWord = dm->dmDefaultSource;
+                oldWord = dm->u.s.dmDefaultSource;
         }
     }
 
@@ -710,9 +710,9 @@
                                       Sel, 0);
         if (dm) {
             if (nIDComboBox == cmb2)
-                dm->u1.s1.dmPaperSize = oldWord;
+                dm->u.s.dmPaperSize = oldWord;
             else
-                dm->dmDefaultSource = oldWord;
+                dm->u.s.dmDefaultSource = oldWord;
         }
     }
     else {
@@ -722,9 +722,9 @@
          */
         if (dm) {
             if (nIDComboBox == cmb2)
-                oldWord = dm->u1.s1.dmPaperSize;
+                oldWord = dm->u.s.dmPaperSize;
             else
-                oldWord = dm->dmDefaultSource;
+                oldWord = dm->u.s.dmDefaultSource;
         }
     }
 
@@ -974,7 +974,7 @@
 	  if (lppd->hDevMode == 0)
 	      copies = lppd->nCopies;
 	  else
-	      copies = lpdm->dmCopies;
+	      copies = lpdm->u.s.dmCopies;
 	  if(copies == 0) copies = 1;
 	  else if(copies < 0) copies = MAX_COPIES;
 	  SetDlgItemInt(hDlg, edt3, copies, FALSE);
@@ -996,7 +996,7 @@
             ShowWindow(GetDlgItem(hDlg, chx1), SW_HIDE);
 
     } else { /* PD_PRINTSETUP */
-      BOOL bPortrait = (lpdm->u1.s1.dmOrientation == DMORIENT_PORTRAIT);
+      BOOL bPortrait = (lpdm->u.s.dmOrientation == DMORIENT_PORTRAIT);
 
       PRINTDLG_SetUpPaperComboBoxA(hDlg, cmb2,
 				  PrintStructures->lpPrinterInfo->pPrinterName,
@@ -1123,7 +1123,7 @@
 	  if (lppd->hDevMode == 0)
 	      copies = lppd->nCopies;
 	  else
-	      copies = lpdm->dmCopies;
+	      copies = lpdm->u.s.dmCopies;
 	  if(copies == 0) copies = 1;
 	  else if(copies < 0) copies = MAX_COPIES;
 	  SetDlgItemInt(hDlg, edt3, copies, FALSE);
@@ -1145,7 +1145,7 @@
             ShowWindow(GetDlgItem(hDlg, chx1), SW_HIDE);
 
     } else { /* PD_PRINTSETUP */
-      BOOL bPortrait = (lpdm->u1.s1.dmOrientation == DMORIENT_PORTRAIT);
+      BOOL bPortrait = (lpdm->u.s.dmOrientation == DMORIENT_PORTRAIT);
 
       PRINTDLG_SetUpPaperComboBoxW(hDlg, cmb2,
 				  PrintStructures->lpPrinterInfo->pPrinterName,
@@ -1519,7 +1519,7 @@
     case rad1: /* Paperorientation */
         if (lppd->Flags & PD_PRINTSETUP)
         {
-              lpdm->u1.s1.dmOrientation = DMORIENT_PORTRAIT;
+              lpdm->u.s.dmOrientation = DMORIENT_PORTRAIT;
               SendDlgItemMessageA(hDlg, ico1, STM_SETIMAGE, (WPARAM) IMAGE_ICON,
                           (LPARAM)(PrintStructures->hPortraitIcon));
         }
@@ -1528,7 +1528,7 @@
     case rad2: /* Paperorientation */
         if (lppd->Flags & PD_PRINTSETUP)
         {
-              lpdm->u1.s1.dmOrientation = DMORIENT_LANDSCAPE;
+              lpdm->u.s.dmOrientation = DMORIENT_LANDSCAPE;
               SendDlgItemMessageA(hDlg, ico1, STM_SETIMAGE, (WPARAM) IMAGE_ICON,
                           (LPARAM)(PrintStructures->hLandscapeIcon));
         }
@@ -1552,7 +1552,7 @@
       {
 	  DWORD Sel = SendDlgItemMessageA(hDlg, cmb2, CB_GETCURSEL, 0, 0);
 	  if(Sel != CB_ERR)
-	      lpdm->u1.s1.dmPaperSize = SendDlgItemMessageA(hDlg, cmb2,
+	      lpdm->u.s.dmPaperSize = SendDlgItemMessageA(hDlg, cmb2,
 							    CB_GETITEMDATA,
 							    Sel, 0);
       }
@@ -1562,7 +1562,7 @@
       {
 	  DWORD Sel = SendDlgItemMessageA(hDlg, cmb3, CB_GETCURSEL, 0, 0);
 	  if(Sel != CB_ERR)
-	      lpdm->dmDefaultSource = SendDlgItemMessageA(hDlg, cmb3,
+	      lpdm->u.s.dmDefaultSource = SendDlgItemMessageA(hDlg, cmb3,
 							  CB_GETITEMDATA, Sel,
 							  0);
       }
@@ -1573,8 +1573,8 @@
 	case rad1:                         /* orientation */
 	case rad2:
 	    if (IsDlgButtonChecked(hDlg, rad1) == BST_CHECKED) {
-	        if(lpdm->u1.s1.dmOrientation != DMORIENT_PORTRAIT) {
-		    lpdm->u1.s1.dmOrientation = DMORIENT_PORTRAIT;
+	        if(lpdm->u.s.dmOrientation != DMORIENT_PORTRAIT) {
+		    lpdm->u.s.dmOrientation = DMORIENT_PORTRAIT;
 		    SendDlgItemMessageA(hDlg, stc10, STM_SETIMAGE,
 					(WPARAM)IMAGE_ICON,
 					(LPARAM)PrintStructures->hPortraitIcon);
@@ -1583,8 +1583,8 @@
 					(LPARAM)PrintStructures->hPortraitIcon);
 		}
 	    } else {
-	        if(lpdm->u1.s1.dmOrientation != DMORIENT_LANDSCAPE) {
-	            lpdm->u1.s1.dmOrientation = DMORIENT_LANDSCAPE;
+	        if(lpdm->u.s.dmOrientation != DMORIENT_LANDSCAPE) {
+	            lpdm->u.s.dmOrientation = DMORIENT_LANDSCAPE;
 		    SendDlgItemMessageA(hDlg, stc10, STM_SETIMAGE,
 					(WPARAM)IMAGE_ICON,
 					(LPARAM)PrintStructures->hLandscapeIcon);
@@ -1683,7 +1683,7 @@
     case rad1: /* Paperorientation */
         if (lppd->Flags & PD_PRINTSETUP)
         {
-              lpdm->u1.s1.dmOrientation = DMORIENT_PORTRAIT;
+              lpdm->u.s.dmOrientation = DMORIENT_PORTRAIT;
               SendDlgItemMessageW(hDlg, ico1, STM_SETIMAGE, (WPARAM) IMAGE_ICON,
                           (LPARAM)(PrintStructures->hPortraitIcon));
         }
@@ -1692,7 +1692,7 @@
     case rad2: /* Paperorientation */
         if (lppd->Flags & PD_PRINTSETUP)
         {
-              lpdm->u1.s1.dmOrientation = DMORIENT_LANDSCAPE;
+              lpdm->u.s.dmOrientation = DMORIENT_LANDSCAPE;
               SendDlgItemMessageW(hDlg, ico1, STM_SETIMAGE, (WPARAM) IMAGE_ICON,
                           (LPARAM)(PrintStructures->hLandscapeIcon));
         }
@@ -1716,7 +1716,7 @@
       {
 	  DWORD Sel = SendDlgItemMessageW(hDlg, cmb2, CB_GETCURSEL, 0, 0);
 	  if(Sel != CB_ERR)
-	      lpdm->u1.s1.dmPaperSize = SendDlgItemMessageW(hDlg, cmb2,
+	      lpdm->u.s.dmPaperSize = SendDlgItemMessageW(hDlg, cmb2,
 							    CB_GETITEMDATA,
 							    Sel, 0);
       }
@@ -1726,7 +1726,7 @@
       {
 	  DWORD Sel = SendDlgItemMessageW(hDlg, cmb3, CB_GETCURSEL, 0, 0);
 	  if(Sel != CB_ERR)
-	      lpdm->dmDefaultSource = SendDlgItemMessageW(hDlg, cmb3,
+	      lpdm->u.s.dmDefaultSource = SendDlgItemMessageW(hDlg, cmb3,
 							  CB_GETITEMDATA, Sel,
 							  0);
       }
@@ -1737,8 +1737,8 @@
 	case rad1:                         /* orientation */
 	case rad2:
 	    if (IsDlgButtonChecked(hDlg, rad1) == BST_CHECKED) {
-	        if(lpdm->u1.s1.dmOrientation != DMORIENT_PORTRAIT) {
-		    lpdm->u1.s1.dmOrientation = DMORIENT_PORTRAIT;
+	        if(lpdm->u.s.dmOrientation != DMORIENT_PORTRAIT) {
+		    lpdm->u.s.dmOrientation = DMORIENT_PORTRAIT;
 		    SendDlgItemMessageW(hDlg, stc10, STM_SETIMAGE,
 					(WPARAM)IMAGE_ICON,
 					(LPARAM)PrintStructures->hPortraitIcon);
@@ -1747,8 +1747,8 @@
 					(LPARAM)PrintStructures->hPortraitIcon);
 		}
 	    } else {
-	        if(lpdm->u1.s1.dmOrientation != DMORIENT_LANDSCAPE) {
-	            lpdm->u1.s1.dmOrientation = DMORIENT_LANDSCAPE;
+	        if(lpdm->u.s.dmOrientation != DMORIENT_LANDSCAPE) {
+	            lpdm->u.s.dmOrientation = DMORIENT_LANDSCAPE;
 		    SendDlgItemMessageW(hDlg, stc10, STM_SETIMAGE,
 					(WPARAM)IMAGE_ICON,
 					(LPARAM)PrintStructures->hLandscapeIcon);
@@ -2513,15 +2513,15 @@
 
     /* Save paper orientation into device context */
     if(pda->curdlg.ptPaperSize.x > pda->curdlg.ptPaperSize.y)
-        dm->u1.s1.dmOrientation = DMORIENT_LANDSCAPE;
+        dm->u.s.dmOrientation = DMORIENT_LANDSCAPE;
     else
-        dm->u1.s1.dmOrientation = DMORIENT_PORTRAIT;
+        dm->u.s.dmOrientation = DMORIENT_PORTRAIT;
 
     /* Save paper size into the device context */
     paperword = SendDlgItemMessageA(hDlg,cmb2,CB_GETITEMDATA,
         SendDlgItemMessageA(hDlg, cmb2, CB_GETCURSEL, 0, 0), 0);
     if (paperword != CB_ERR)
-        dm->u1.s1.dmPaperSize = paperword;
+        dm->u.s.dmPaperSize = paperword;
     else
         FIXME("could not get dialog text for papersize cmbbox?\n");
 
@@ -2529,7 +2529,7 @@
     paperword = SendDlgItemMessageA(hDlg,cmb1,CB_GETITEMDATA,
         SendDlgItemMessageA(hDlg, cmb1, CB_GETCURSEL, 0, 0), 0);
     if (paperword != CB_ERR)
-        dm->dmDefaultSource = paperword;
+        dm->u.s.dmDefaultSource = paperword;
     else
         FIXME("could not get dialog text for papersize cmbbox?\n");
 
@@ -2580,9 +2580,9 @@
 
     /* Save orientation */
     if (pda->dlga->ptPaperSize.x > pda->dlga->ptPaperSize.y)
-        dm->u1.s1.dmOrientation = DMORIENT_LANDSCAPE;
+        dm->u.s.dmOrientation = DMORIENT_LANDSCAPE;
     else
-        dm->u1.s1.dmOrientation = DMORIENT_PORTRAIT;
+        dm->u.s.dmOrientation = DMORIENT_PORTRAIT;
 
     GlobalUnlock(pda->pdlg.hDevNames);
     GlobalUnlock(pda->pdlg.hDevMode);
@@ -2834,7 +2834,7 @@
     case cmb3:
 	if(msg == CBN_SELCHANGE){
 	    DEVMODEA *dm = GlobalLock(pda->pdlg.hDevMode);
-	    dm->dmDefaultSource = SendDlgItemMessageA(hDlg, cmb3,CB_GETITEMDATA,
+	    dm->u.s.dmDefaultSource = SendDlgItemMessageA(hDlg, cmb3,CB_GETITEMDATA,
                 SendDlgItemMessageA(hDlg, cmb3, CB_GETCURSEL, 0, 0), 0);
 	    GlobalUnlock(pda->pdlg.hDevMode);
 	}
@@ -2857,10 +2857,10 @@
 	                        DM_IN_BUFFER | DM_OUT_BUFFER | DM_IN_PROMPT);
 	    ClosePrinter(hPrinter);
 	    /* Changing paper */
-	    PRINTDLG_PaperSizeA(&(pda->pdlg), dm->u1.s1.dmPaperSize, &(pda->curdlg.ptPaperSize));
+	    PRINTDLG_PaperSizeA(&(pda->pdlg), dm->u.s.dmPaperSize, &(pda->curdlg.ptPaperSize));
 	    pda->curdlg.ptPaperSize.x = _c_10mm2size(pda->dlga, pda->curdlg.ptPaperSize.x);
 	    pda->curdlg.ptPaperSize.y = _c_10mm2size(pda->dlga, pda->curdlg.ptPaperSize.y);
-            if (dm->u1.s1.dmOrientation == DMORIENT_LANDSCAPE){
+            if (dm->u.s.dmOrientation == DMORIENT_LANDSCAPE){
                 DWORD tmp = pda->curdlg.ptPaperSize.x;
                 pda->curdlg.ptPaperSize.x = pda->curdlg.ptPaperSize.y;
                 pda->curdlg.ptPaperSize.y = tmp;
@@ -2874,7 +2874,7 @@
 	    count = SendDlgItemMessageA(hDlg, cmb2, CB_GETCOUNT, 0, 0);
 	    if(count != CB_ERR){ 
                 for(i=0; i<count; ++i){
-		    if(SendDlgItemMessageA(hDlg, cmb2, CB_GETITEMDATA, i, 0) == dm->u1.s1.dmPaperSize) {
+		    if(SendDlgItemMessageA(hDlg, cmb2, CB_GETITEMDATA, i, 0) == dm->u.s.dmPaperSize) {
 			SendDlgItemMessageA(hDlg, cmb2, CB_SETCURSEL, i, 0);
 			break;
 		    }
@@ -3180,7 +3180,7 @@
         if(pda->dlga->hDevMode)
         {
             dm = GlobalLock(pda->dlga->hDevMode);
-            if (dm->u1.s1.dmOrientation == DMORIENT_LANDSCAPE)
+            if (dm->u.s.dmOrientation == DMORIENT_LANDSCAPE)
                 CheckRadioButton(hDlg, rad1, rad2, rad2);
             else /* this is default if papersize is not set */
                 CheckRadioButton(hDlg, rad1, rad2, rad1);
@@ -3226,8 +3226,8 @@
         PRINTDLG_PS_ChangePrinterA(hDlg, pda);
 	dm = GlobalLock(pda->pdlg.hDevMode);
 	if(dm){
-	    dm->dmDefaultSource = 15; /*FIXME: Automatic select. Does it always 15 at start? */
-	    PRINTDLG_PaperSizeA(&(pda->pdlg), dm->u1.s1.dmPaperSize, &pda->curdlg.ptPaperSize);
+	    dm->u.s.dmDefaultSource = 15; /*FIXME: Automatic select. Does it always 15 at start? */
+	    PRINTDLG_PaperSizeA(&(pda->pdlg), dm->u.s.dmPaperSize, &pda->curdlg.ptPaperSize);
             GlobalUnlock(pda->pdlg.hDevMode);
 	    pda->curdlg.ptPaperSize.x = _c_10mm2size(pda->dlga, pda->curdlg.ptPaperSize.x);
 	    pda->curdlg.ptPaperSize.y = _c_10mm2size(pda->dlga, pda->curdlg.ptPaperSize.y);
@@ -3297,7 +3297,7 @@
         {
             dm = GlobalLock(pda->dlga->hDevMode);
             /* Landscape orientation */
-            if (dm->u1.s1.dmOrientation == DMORIENT_LANDSCAPE)
+            if (dm->u.s.dmOrientation == DMORIENT_LANDSCAPE)
                 CheckRadioButton(hDlg, rad1, rad2, rad2);
             else /* this is default if papersize is not set */
                 CheckRadioButton(hDlg, rad1, rad2, rad1);
@@ -3439,7 +3439,7 @@
 	setupdlg->hDevMode = pdlg.hDevMode;
 	setupdlg->hDevNames = pdlg.hDevNames;
         dm = GlobalLock(pdlg.hDevMode);
-        PRINTDLG_PaperSizeA(&pdlg, dm->u1.s1.dmPaperSize, &setupdlg->ptPaperSize);
+        PRINTDLG_PaperSizeA(&pdlg, dm->u.s.dmPaperSize, &setupdlg->ptPaperSize);
         GlobalUnlock(pdlg.hDevMode);
         setupdlg->ptPaperSize.x=_c_10mm2size(setupdlg,setupdlg->ptPaperSize.x);
         setupdlg->ptPaperSize.y=_c_10mm2size(setupdlg,setupdlg->ptPaperSize.y);
