Index: shell32_En.rc
===================================================================
RCS file: /home/wine/wine/dlls/shell32/shell32_En.rc,v
retrieving revision 1.14
diff -u -r1.14 shell32_En.rc
--- shell32_En.rc	9 Jul 2004 22:51:19 -0000	1.14
+++ shell32_En.rc	12 Jul 2004 21:01:03 -0000
@@ -181,8 +181,8 @@
 /* message box strings */
 STRINGTABLE DISCARDABLE
 {
-       IDS_RESTART_TITLE       "Restart"
-       IDS_RESTART_PROMPT      "Do you want to simulate a Windows reboot?"
-       IDS_SHUTDOWN_TITLE      "Shutdown"
-       IDS_SHUTDOWN_PROMPT     "Do you want to shutdown your Wine session?"
+	IDS_RESTART_TITLE	"Restart"
+	IDS_RESTART_PROMPT	"Do you want to restart the system?"
+	IDS_SHUTDOWN_TITLE	"Shutdown"
+	IDS_SHUTDOWN_PROMPT	"Do you want to shutdown?"
 }
Index: shellpath.c
===================================================================
RCS file: /home/wine/wine/dlls/shell32/shellpath.c,v
retrieving revision 1.87
diff -u -r1.87 shellpath.c
--- shellpath.c	20 Aug 2004 19:57:50 -0000	1.87
+++ shellpath.c	20 Sep 2004 18:20:35 -0000
@@ -1485,6 +1485,10 @@
             *ppidl = _ILCreatePrinters();
             break;
 
+        case CSIDL_FONTS:
+            FIXME("virtual font folder");
+            break;
+
         case CSIDL_BITBUCKET:
             *ppidl = _ILCreateBitBucket();
             break;
Index: shell32_main.c
===================================================================
RCS file: /home/wine/wine/dlls/shell32/shell32_main.c,v
retrieving revision 1.136
diff -u -r1.136 shell32_main.c
--- shell32_main.c	14 Sep 2004 17:44:40 -0000	1.136
+++ shell32_main.c	21 Sep 2004 19:14:29 -0000
@@ -953,7 +953,6 @@
 	    InitCommonControlsEx(NULL);
 
 	    SIC_Initialize();
-	    SYSTRAY_Init();
 	    InitChangeNotifications();
 	    break;
 
Index: shell32_main.h
===================================================================
RCS file: /home/wine/wine/dlls/shell32/shell32_main.h,v
retrieving revision 1.91
diff -u -r1.91 shell32_main.h
--- shell32_main.h	22 Aug 2004 21:38:46 -0000	1.91
+++ shell32_main.h	21 Sep 2004 19:24:47 -0000
@@ -128,10 +128,6 @@
 	(((kst) & MK_SHIFT) ? DROPEFFECT_LINK : DROPEFFECT_COPY):\
 	DROPEFFECT_MOVE)
 
-/* Systray */
-BOOL SYSTRAY_Init(void);
-
-
 HGLOBAL RenderHDROP(LPITEMIDLIST pidlRoot, LPITEMIDLIST * apidl, UINT cidl);
 HGLOBAL RenderSHELLIDLIST (LPITEMIDLIST pidlRoot, LPITEMIDLIST * apidl, UINT cidl);
 HGLOBAL RenderSHELLIDLISTOFFSET (LPITEMIDLIST pidlRoot, LPITEMIDLIST * apidl, UINT cidl);
