Index: marshal.c
===================================================================
--- marshal.c	(revision 38899)
+++ marshal.c	(working copy)
@@ -824,7 +824,7 @@
     } while (new_dest_context != old_dest_context);
 
     if (dest_context_data)
-        InterlockedExchangePointer(&This->dest_context_data, dest_context_data);
+        (void)InterlockedExchangePointer(&This->dest_context_data, dest_context_data);
 }
 
 static HRESULT proxy_manager_query_local_interface(struct proxy_manager * This, REFIID riid, void ** ppv)
