Index: Makefile.in
===================================================================
RCS file: /home/wine/wine/dlls/comctl32/Makefile.in,v
retrieving revision 1.32
diff -u -r1.32 Makefile.in
--- Makefile.in	11 Oct 2003 01:09:20 -0000	1.32
+++ Makefile.in	22 Jan 2004 18:56:34 -0000
@@ -41,9 +41,14 @@
 RC_BINARIES = \
 	idb_hist_large.bmp \
 	idb_hist_small.bmp \
+	idb_std_small.bmp \
 	idb_std_large.bmp \
 	idb_view_large.bmp \
-	idb_view_small.bmp 
+	idb_view_small.bmp \
+	idc_divider.cur \
+	idc_divideropen.cur \
+	idi_dragarrow.ico \
+	idt_check.bmp
 
 SUBDIRS = tests
 
Index: comctl32undoc.c
===================================================================
RCS file: /home/wine/wine/dlls/comctl32/comctl32undoc.c,v
retrieving revision 1.86
diff -u -r1.86 comctl32undoc.c
--- comctl32undoc.c	20 Nov 2003 04:19:41 -0000	1.86
+++ comctl32undoc.c	22 Jan 2004 18:56:35 -0000
@@ -1669,21 +1669,7 @@
 HDPA WINAPI
 DPA_Create (INT nGrow)
 {
-    HDPA hdpa;
-
-    TRACE("(%d)\n", nGrow);
-
-    hdpa = Alloc (sizeof(*hdpa));
-    if (hdpa) {
-	hdpa->nGrow = max(8, nGrow);
-	hdpa->hHeap = COMCTL32_hHeap;
-	hdpa->nMaxCount = hdpa->nGrow * 2;
-	hdpa->ptrs = Alloc (hdpa->nMaxCount * sizeof(LPVOID));
-    }
-
-    TRACE("-- %p\n", hdpa);
-
-    return hdpa;
+    return DPA_CreateEx (nGrow, GetProcessHeap());
 }
 
 
Index: rsrc.rc
===================================================================
RCS file: /home/wine/wine/dlls/comctl32/rsrc.rc,v
retrieving revision 1.36
diff -u -r1.36 rsrc.rc
--- rsrc.rc	12 Dec 2003 05:57:22 -0000	1.36
+++ rsrc.rc	22 Jan 2004 18:56:36 -0000
@@ -29,8 +29,9 @@
 
 #include "wine/wine_common_ver.rc"
 
-IDT_CHECK BITMAP LOADONCALL DISCARDABLE
-{
+/* BINRES idt_check.bmp */
+IDT_CHECK BITMAP LOADONCALL DISCARDABLE idt_check.bmp
+/* {
  '42 4D 8E 01 00 00 00 00 00 00 5E 00 00 00 28 00'
  '00 00 1F 00 00 00 10 00 00 00 01 00 04 00 02 00'
  '00 00 30 01 00 00 01 00 00 00 01 00 00 00 0A 00'
@@ -56,10 +57,11 @@
  '08 99 01 00 01 99 01 00 01 11 01 99 00 00 01 99'
  '0D 11 02 99 0A 11 02 00 02 11 01 99 00 00 01 99'
  '0D 11 02 99 0E 11 01 99 00 00 1F 99 00 01'
-}
+} */
 
-IDB_STD_SMALL BITMAP LOADONCALL DISCARDABLE
-{
+/* BINRES idb_std_small.bmp */
+IDB_STD_SMALL BITMAP LOADONCALL DISCARDABLE idb_std_small.bmp
+/* {
  '42 4D 20 07 00 00 00 00 00 00 76 00 00 00 28 00'
  '00 00 F0 00 00 00 10 00 00 00 01 00 04 00 02 00'
  '00 00 AA 06 00 00 00 00 00 00 00 00 00 00 00 00'
@@ -174,7 +176,7 @@
  '06 88 02 84 04 44 04 88 02 66 04 88 02 66 12 88'
  '02 11 06 88 02 86 08 66 02 88 00 00 90 88 08 66'
  '02 68 2A 88 04 66 14 88 02 18 12 88 00 00 00 01'
-}
+} */
 
 /* BINRES idb_std_large.bmp */
 IDB_STD_LARGE BITMAP LOADONCALL DISCARDABLE idb_std_large.bmp
@@ -969,8 +971,9 @@
 } */
 
 
-IDC_DIVIDER CURSOR LOADONCALL DISCARDABLE
-{
+/* BINRES idc_divider.cur */
+IDC_DIVIDER CURSOR LOADONCALL DISCARDABLE idc_divider.cur
+/* {
  '00 00 02 00 01 00 20 20 00 00 09 00 09 00 30 01'
  '00 00 16 00 00 00 28 00 00 00 20 00 00 00 40 00'
  '00 00 01 00 01 00 00 00 00 00 80 00 00 00 00 00'
@@ -992,11 +995,12 @@
  '3F FF 00 00 3F FF 80 00 7F FF C6 18 FF FF E6 19'
  'FF FF FE 1F FF FF FE 1F FF FF FE 1F FF FF FE 1F'
  'FF FF FE 1F FF FF'
-}
+} */
 
 
-IDC_DIVIDEROPEN CURSOR LOADONCALL DISCARDABLE
-{
+/* BINRES idc_divideropen.cur */
+IDC_DIVIDEROPEN CURSOR LOADONCALL DISCARDABLE idc_divideropen.cur
+/* {
  '00 00 02 00 01 00 20 20 00 00 09 00 09 00 30 01'
  '00 00 16 00 00 00 28 00 00 00 20 00 00 00 40 00'
  '00 00 01 00 01 00 00 00 00 00 80 00 00 00 00 00'
@@ -1018,11 +1022,12 @@
  '3F FF 00 00 3F FF 80 00 7F FF C4 08 FF FF E4 09'
  'FF FF FC 0F FF FF FC 0F FF FF FC 0F FF FF FC 0F'
  'FF FF FC 0F FF FF'
-}
+} */
 
 
-IDI_DRAGARROW ICON LOADONCALL DISCARDABLE
-{
+/* BINRES idi_dragarrow.ico */
+IDI_DRAGARROW ICON LOADONCALL DISCARDABLE idi_dragarrow.ico
+/* {
  '00 00 01 00 01 00 20 20 02 00 00 00 00 00 30 01'
  '00 00 16 00 00 00 28 00 00 00 20 00 00 00 40 00'
  '00 00 01 00 01 00 00 00 00 00 00 01 00 00 00 00'
@@ -1044,7 +1049,7 @@
  'FF FF F0 03 FF FF F0 01 FF FF F0 03 FF FF F0 07'
  'FF FF F0 8F FF FF F0 9F FF FF FF FF FF FF FF FF'
  'FF FF FF FF FF FF'
-}
+} */
 
 
 /*
Index: toolbar.c
===================================================================
RCS file: /home/wine/wine/dlls/comctl32/toolbar.c,v
retrieving revision 1.147
diff -u -r1.147 toolbar.c
--- toolbar.c	26 Nov 2003 22:05:02 -0000	1.147
+++ toolbar.c	22 Jan 2004 18:56:39 -0000
@@ -5042,6 +5042,7 @@
     if((infoPtr->nHotItem >= 0) && (nHit != -1))
         infoPtr->buttons[infoPtr->nHotItem].bHot = TRUE;
 
+    if (0 <= infoPtr->nButtonDown) {
 	btnPtr = &infoPtr->buttons[infoPtr->nButtonDown];
 	btnPtr->fsState &= ~TBSTATE_PRESSED;
 
@@ -5111,6 +5112,7 @@
 	    nmmouse.dwItemData = btnPtr->dwData;
 	    TOOLBAR_SendNotify ((NMHDR *) &nmmouse, infoPtr, NM_CLICK);
 	}
+    }
     return 0;
 }
 
