diff --git a/dll/directx/wine/wined3d/surface.c b/dll/directx/wine/wined3d/surface.c
index 7ebb765..3593455 100644
--- a/dll/directx/wine/wined3d/surface.c
+++ b/dll/directx/wine/wined3d/surface.c
@@ -163,16 +163,16 @@ void texture2d_blt_fbo(struct wined3d_device *device, struct wined3d_context *co
 
     switch (filter)
     {
-        case WINED3D_TEXF_LINEAR:
-            gl_filter = GL_LINEAR;
+        case WINED3D_TEXF_NONE:
+        case WINED3D_TEXF_POINT:
+            gl_filter = GL_NEAREST;
             break;
 
         default:
             FIXME("Unsupported filter mode %s (%#x).\n", debug_d3dtexturefiltertype(filter), filter);
             /* fall through */
-        case WINED3D_TEXF_NONE:
-        case WINED3D_TEXF_POINT:
-            gl_filter = GL_NEAREST;
+        case WINED3D_TEXF_LINEAR:
+            gl_filter = GL_LINEAR;
             break;
     }
 
