Index: rpcss_main.c
===================================================================
--- rpcss_main.c	(revision 37240)
+++ rpcss_main.c	(working copy)
@@ -62,7 +62,7 @@

 //extern HANDLE __wine_make_process_system(void);

-static BOOL RPCSS_Initialize(void)
+BOOL RPCSS_Initialize(void)
 {
   static unsigned short irot_protseq[] = IROT_PROTSEQ;
   static unsigned short irot_endpoint[] = IROT_ENDPOINT;
@@ -116,7 +116,7 @@

 /* returns false if we discover at the last moment that we
    aren't ready to terminate */
-static BOOL RPCSS_Shutdown(void)
+BOOL RPCSS_Shutdown(void)
 {
   RpcMgmtStopServerListening(NULL);
   RpcServerUnregisterIf(epm_v3_0_s_ifspec, NULL, TRUE);
@@ -127,6 +127,7 @@
   return TRUE;
 }

+#if 0
 int main( int argc, char **argv )
 {
   /*
@@ -142,3 +143,4 @@

   return 0;
 }
+#endif
