Index: string.c
===================================================================
RCS file: /home/wine/wine/libs/unicode/string.c,v
retrieving revision 1.5
diff -u -r1.5 string.c
--- string.c	2 Apr 2004 19:40:31 -0000	1.5
+++ string.c	16 Apr 2004 09:07:08 -0000
@@ -23,6 +23,21 @@
 
 #include "wine/unicode.h"
 
+#undef strlenW
+#undef strcpyW
+#undef strcatW
+#undef strstrW
+#undef strtolW
+#undef strchrW
+#undef strrchrW
+#undef strncmpW
+#undef strncpyW
+#undef strcmpW
+#undef strcmpiW
+#undef strncmpiW
+#undef sprintfW
+#undef snprintfW
+
 int strcmpiW( const WCHAR *str1, const WCHAR *str2 )
 {
     for (;;)
