Index: cabextract.c
===================================================================
RCS file: /home/wine/wine/dlls/cabinet/cabextract.c,v
retrieving revision 1.15
diff -u -r1.15 cabextract.c
--- cabextract.c	30 Nov 2004 21:39:00 -0000	1.15
+++ cabextract.c	5 Dec 2004 21:40:05 -0000
@@ -2515,7 +2515,7 @@
       cabname = (fi->folder->cab[0]->filename);
     }
 
-    ERR(errmsg, cabname);
+    ERR((char *)errmsg, cabname);
   }
 }
 
