Index: ifs.h
===================================================================
RCS file: /home/wine/wine/dlls/ole32/ifs.h,v
retrieving revision 1.11
diff -u -r1.11 ifs.h
--- ifs.h	23 Jan 2004 22:51:42 -0000	1.11
+++ ifs.h	28 Jan 2004 21:32:28 -0000
@@ -35,6 +35,7 @@
 
 typedef struct IMalloc16 IMalloc16, *LPMALLOC16;
 
+#undef INTERFACE
 #define INTERFACE IMalloc16
 #define IMalloc16_METHODS \
     IUnknown_METHODS \
