Index: Makefile.in
===================================================================
RCS file: /home/wine/wine/dlls/ole32/Makefile.in,v
retrieving revision 1.39
diff -u -r1.39 Makefile.in
--- Makefile.in	21 Sep 2004 00:35:03 -0000	1.39
+++ Makefile.in	20 Oct 2004 09:21:24 -0000
@@ -53,7 +53,7 @@
 	ole2thk.spec \
 	storage.spec
 
-RC_SRCS = ole32res.rc version.rc
+RC_SRCS = ole32res.rc
 RC_BINSRC = ole32res.rc
 RC_BINARIES = \
 	drag_copy.cur \
Index: ifs.h
===================================================================
RCS file: /home/wine/wine/dlls/ole32/ifs.h,v
retrieving revision 1.13
diff -u -r1.13 ifs.h
--- ifs.h	5 Oct 2004 04:16:21 -0000	1.13
+++ ifs.h	20 Oct 2004 09:21:25 -0000
@@ -33,8 +33,7 @@
  * IMalloc16 interface
  */
 
-typedef struct IMalloc16 IMalloc16, *LPMALLOC16;
-
+#undef INTERFACE
 #define INTERFACE IMalloc16
 DECLARE_INTERFACE_(IMalloc16,IUnknown)
 {
@@ -52,14 +51,14 @@
 };
 #undef INTERFACE
 
+typedef struct IMalloc16 *LPMALLOC16;
+
 /**********************************************************************/
 
 extern LPMALLOC16 IMalloc16_Constructor();
 
 /**********************************************************************/
 
-typedef struct ILockBytes16 *LPLOCKBYTES16, ILockBytes16;
-
 #define INTERFACE ILockBytes16
 DECLARE_INTERFACE_(ILockBytes16,IUnknown)
 {
@@ -95,8 +94,6 @@
     DWORD reserved;
 } STATSTG16;
 
-typedef struct IStream16 IStream16, *LPSTREAM16;
-
 #define INTERFACE IStream16
 DECLARE_INTERFACE_(IStream16,ISequentialStream)
 {
@@ -123,8 +120,6 @@
 /**********************************************************************/
 
 typedef OLECHAR16 **SNB16;
-
-typedef struct IStorage16 IStorage16, *LPSTORAGE16;
 
 #define INTERFACE IStorage16
 DECLARE_INTERFACE_(IStorage16,IUnknown)
Index: ole32res.rc
===================================================================
RCS file: /home/wine/wine/dlls/ole32/ole32res.rc,v
retrieving revision 1.5
diff -u -r1.5 ole32res.rc
--- ole32res.rc	3 Oct 2003 05:01:34 -0000	1.5
+++ ole32res.rc	20 Oct 2004 09:21:25 -0000
@@ -23,6 +23,8 @@
 #include "winuser.h"
 #include "winnls.h"
 
+#include "version.rc"
+
 /*
  * Everything that does not depend on language,
  * like textless bitmaps etc, go into the
Index: oleproxy.c
===================================================================
RCS file: /home/wine/wine/dlls/ole32/oleproxy.c,v
retrieving revision 1.23
diff -u -r1.23 oleproxy.c
--- oleproxy.c	7 Oct 2004 03:06:49 -0000	1.23
+++ oleproxy.c	20 Oct 2004 09:21:25 -0000
@@ -38,6 +38,7 @@
 
 #include <stdlib.h>
 #include <stdarg.h>
+#include <stddef.h>
 #include <stdio.h>
 #include <string.h>
 
