Index: preproc.c
===================================================================
--- preproc.c	(revision 28293)
+++ preproc.c	(revision 28294)
@@ -328,6 +328,15 @@
 	char *tok;
 	char *cpy = pp_xstrdup(path);
 
+	/* check for absolute windows paths */
+	if (strchr(cpy, ':') != NULL)
+	{
+		nincludepath++;
+		includepath = pp_xrealloc(includepath, nincludepath * sizeof(*includepath));
+		includepath[nincludepath-1] = cpy;
+		return;
+	}
+
 	tok = strtok(cpy, INCLUDESEPARATOR);
 	while(tok)
 	{
