diff -pudN e:\wine\dlls\comdlg32/cdlg32.c e:\reactos\dll\win32\comdlg32/cdlg32.c
--- e:\wine\dlls\comdlg32/cdlg32.c	2016-11-16 17:28:42 +0100
+++ e:\reactos\dll\win32\comdlg32/cdlg32.c	2015-08-27 22:02:19 +0100
@@ -157,6 +157,8 @@ DWORD WINAPI CommDlgExtendedError(void)
 	  return 0; /* we never set an error, so there isn't one */
 }
 
+#ifndef __REACTOS__ /* Win 7 */
+
 /*************************************************************************
  * Implement the CommDlg32 class factory
  *
@@ -266,7 +268,11 @@ HRESULT WINAPI DllGetClassObject(REFCLSI
  */
 HRESULT WINAPI DllRegisterServer(void)
 {
+#ifdef __REACTOS__
+    return E_FAIL; // FIXME: __wine_register_resources(COMDLG32_hInstance);
+#else
     return __wine_register_resources(COMDLG32_hInstance);
+#endif
 }
 
 /***********************************************************************
@@ -274,5 +280,11 @@ HRESULT WINAPI DllRegisterServer(void)
  */
 HRESULT WINAPI DllUnregisterServer(void)
 {
+#ifdef __REACTOS__
+    return E_FAIL; // FIXME: __wine_unregister_resources(COMDLG32_hInstance);
+#else
     return __wine_unregister_resources(COMDLG32_hInstance);
+#endif
 }
+
+#endif /* Win 7 */
diff -pudN e:\wine\dlls\comdlg32/itemdlg.c e:\reactos\dll\win32\comdlg32/itemdlg.c
--- e:\wine\dlls\comdlg32/itemdlg.c	2016-08-14 19:11:13 +0100
+++ e:\reactos\dll\win32\comdlg32/itemdlg.c	2016-08-15 16:16:18 +0100
@@ -18,6 +18,8 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
+#ifndef __REACTOS__ /* Win 7 */
+
 #include <stdarg.h>
 
 #define COBJMACROS
@@ -4659,3 +4661,5 @@ HRESULT FileSaveDialog_Constructor(IUnkn
 {
     return FileDialog_constructor(pUnkOuter, riid, ppv, ITEMDLG_TYPE_SAVE);
 }
+
+#endif /* Win 7 */
