diff -prudN e:\Wine\dlls\comdlg32/cdlg32.c e:\reactos\dll\win32\comdlg32/cdlg32.c
--- e:\Wine\dlls\comdlg32/cdlg32.c	2013-10-26 22:24:27.508648000 +0100
+++ e:\reactos\dll\win32\comdlg32/cdlg32.c	2014-03-05 11:46:06.260510700 +0100
@@ -174,6 +157,8 @@ DWORD WINAPI CommDlgExtendedError(void)
 	  return 0; /* we never set an error, so there isn't one */
 }
 
+#ifndef __REACTOS__ /* Win 7 */
+
 /*************************************************************************
  * Implement the CommDlg32 class factory
  *
@@ -283,7 +268,7 @@ HRESULT WINAPI DllGetClassObject(REFCLSI
  */
 HRESULT WINAPI DllRegisterServer(void)
 {
-    return __wine_register_resources(COMDLG32_hInstance);
+    return E_FAIL; // FIXME: __wine_register_resources(COMDLG32_hInstance);
 }
 
 /***********************************************************************
@@ -291,5 +276,7 @@ HRESULT WINAPI DllRegisterServer(void)
  */
 HRESULT WINAPI DllUnregisterServer(void)
 {
-    return __wine_unregister_resources(COMDLG32_hInstance);
+    return E_FAIL; // FIXME: __wine_unregister_resources(COMDLG32_hInstance);
 }
+
+#endif /* Win 7 */

diff -prudN e:\Wine\dlls\comdlg32/comdlg32.spec e:\reactos\dll\win32\comdlg32/comdlg32.spec
--- e:\Wine\dlls\comdlg32/comdlg32.spec	2011-09-16 23:22:36.026770600 +0100
+++ e:\reactos\dll\win32\comdlg32/comdlg32.spec	2012-02-15 22:32:22.456422900 +0100
@@ -3,9 +3,9 @@
 @ stdcall ChooseFontA(ptr)
 @ stdcall ChooseFontW(ptr)
 @ stdcall CommDlgExtendedError()
-@ stdcall -private DllGetClassObject(ptr ptr ptr)
-@ stdcall -private DllRegisterServer()
-@ stdcall -private DllUnregisterServer()
+;@ stdcall -private DllGetClassObject(ptr ptr ptr) ; Win 7
+;@ stdcall -private DllRegisterServer() ; Win 7
+;@ stdcall -private DllUnregisterServer() ; Win 7
 @ stdcall FindTextA(ptr)
 @ stdcall FindTextW(ptr)
 @ stdcall GetFileTitleA(str ptr long)

diff -prudN e:\Wine\dlls\comdlg32/itemdlg.c e:\reactos\dll\win32\comdlg32/itemdlg.c
--- e:\Wine\dlls\comdlg32/itemdlg.c	2013-10-26 22:24:27.515652700 +0100
+++ e:\reactos\dll\win32\comdlg32/itemdlg.c	2013-10-15 19:32:27.803008200 +0100
@@ -18,6 +18,8 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
+#ifndef __REACTOS__ /* Win 7 */
+
 #include <stdarg.h>
 
 #define COBJMACROS
@@ -3660,3 +3662,5 @@ HRESULT FileSaveDialog_Constructor(IUnkn
 {
     return FileDialog_constructor(pUnkOuter, riid, ppv, ITEMDLG_TYPE_SAVE);
 }
+
+#endif /* Win 7 */
